### PR TITLE
virttest.qemu_monitor: Add event "BALLOON_CHANGE" check for balloon operation.

### PR DESCRIPTION
**Pull request by CongLi**
Signed-off-by: Cong Li <coli@redhat.com>

id: 1342726

### CODE REVIEW

**Comment by CongLi** [comment]
11:27:44 INFO | BALLOON_CHANGE QMP event received
11:27:51 INFO | PASS 1-smp_8.2048m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.balloon_check.balloon_evict_and_enlarge.balloon_base

File: virttest/qemu_vm.py [hunk 1]
```diff
@@ -4224,4 +4224,18 @@ def balloon(self, size):
             size = "%s MB" % size
         normalize_data_size = utils_misc.normalize_data_size
         size = int(float(normalize_data_size(size, 'B', '1024')))
-        return self.monitor.balloon(size)
+        # Clear the event list of all QMP monitors
```
Comment (PandaWei) [inline virttest/qemu_vm.py]: IMO,  these new codes should be added in  class QMPMonitor().balloon().
and if so, it's no need to check the vm.protocol.

File: virttest/qemu_vm.py [hunk 2]
```diff
@@ -4224,4 +4224,18 @@ def balloon(self, size):
...
+        qmp_monitors = [m for m in self.monitors if m.protocol == "qmp"]
+        for m in qmp_monitors:
+            m.clear_events()
+        # Send a balloon monitor command
+        self.monitor.balloon(size)
+        # Look for BALLOON QMP events
+        time.sleep(1)
```
Comment (PandaWei) [inline virttest/qemu_vm.py]: Is this mandatory?

File: virttest/qemu_vm.py (hunk 1 above)
Comment (xutian) [inline virttest/qemu_vm.py]: Agree with @PandaWei , event check should be in class QMPMonitor.balloon

**Comment by CongLi** [comment]
@PandaWei @xutian 
Agree with you, and have been updated the code.

Btw, it's better to sleep few secs when check qmp event, as qmp is asynchronous.

**Comment by CongLi** [comment]
1. Error info when failed:
15:04:09 ERROR| QMPEventError: QMP event BALLOON_CHANGE not received after balloon (monitor 'qmpmonitor1')
2. 15:13:24 INFO | PASS 1-smp_8.2048m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.balloon_check.balloon_evict_and_enlarge.balloon_base

File: virttest/qemu_monitor.py [hunk 3]
```diff
@@ -2087,8 +2100,19 @@ def balloon(self, size):
...
-        self.send_args_cmd("balloon value=%s" % size)
+        cmd = "balloon"
+        qmp_event = "BALLOON_CHANGE"
+        self.verify_supported_cmd(cmd)
+        # Clear the event list of QMP monitors
+        self.clear_events()
+        # Send a balloon monitor command
+        self.send_args_cmd("%s value=%s" % (cmd, size))
+        # Look for BALLOON QMP events
+        time.sleep(1)
```
Comment (xutian) [inline virttest/qemu_monitor.py]: better to use wait for  here,  I'm not sure is it enough to wait for 1s to get this event.

File: virttest/qemu_monitor.py [hunk 4]
```diff
@@ -2087,8 +2100,19 @@ def balloon(self, size):
 
         :param size: int type values.
         """
-        self.verify_supported_cmd("balloon")
-        self.send_args_cmd("balloon value=%s" % size)
+        cmd = "balloon"
+        qmp_event = "BALLOON_CHANGE"
+        self.verify_supported_cmd(cmd)
+        # Clear the event list of QMP monitors
+        self.clear_events()
```
Comment (xutian) [inline virttest/qemu_monitor.py]: why not only clear balloon_change events ?

**Comment by CongLi** [comment]
1. Error info when failed:
16:07:43 ERROR| QMPEventError: QMP event BALLOON_CHANGE not received after balloon (monitor 'qmpmonitor1')
2. 16:01:19 INFO | PASS 1-smp_8.2048m.repeat1.run_test.Host_RHEL.m7.u3.qcow2.virtio_blk.up.virtio_net.RHEL.7.2.x86_64.io-github-autotest-qemu.balloon_check.balloon_evict_and_enlarge.balloon_base

@xutian Updated.

File: virttest/qemu_monitor.py [hunk 5]
```diff
@@ -2087,8 +2100,17 @@ def balloon(self, size):
...
+        qmp_event = "BALLOON_CHANGE"
+        self.verify_supported_cmd(cmd)
+        # Clear the event list of QMP monitors
+        self.clear_event(qmp_event)
+        # Send a balloon monitor command
+        self.send_args_cmd("%s value=%s" % (cmd, size))
+        # Look for BALLOON QMP events
+        if not utils_misc.wait_for(lambda: self.get_event(qmp_event), 5):
+            raise QMPEventError(cmd, qmp_event, self.name)
+        logging.info("%s QMP event received" % qmp_event)
```
Comment (xutian) [inline virttest/qemu_monitor.py]: do we really need this line?  if results is expect,  keep quiet is our style.

Comment (CongLi) [inline virttest/qemu_monitor.py]: IMO, it's better to show there is qmp event check in the log,
or users won't know there is event check from log.

**Comment by xutian** [comment]
looks fine, thanks @CongLi . merged

**Pull request merged**